Step3 @MVC 구현(힌트)

juyoungyoo opened the pull request:
안녕하세요 권용근님 
좋은 리뷰 부탁드립니다 :)

juyoungyoo committed:
refactor #step3 view Redirect 구현

juyoungyoo committed:
refactor #step3 리팩토링

juyoungyoo committed:
refactor #step3 controllerScanner 구현 및 리팩토링

kingbbode commented on src/main/java/core/mvc/tobe/HandlerExecution.java:
```diff
@@ -1,10 +1,11 @@
 package core.mvc.tobe;
 
+import com.sun.xml.internal.ws.handler.HandlerException;
```
의도하신건가요? HandleException 오타인가요??

juyoungyoo replied:
오타입니다 수정하였습니다

kingbbode commented on src/main/java/core/mvc/tobe/NotFoundServletException.java:
```diff
@@ -0,0 +1,10 @@
+package core.mvc.tobe;
+
+public class NotFoundServletException extends IllegalArgumentException {
```
NotFound 라면 응답 상태코드는 404가 나와야하지 않을까요? 그에 대한 처리가 보이지 않네요!

kingbbode commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -1,75 +1,66 @@
+            View view = modelAndView.getView();
+            view.render(modelAndView.getModel(), req, resp);
+        } catch (Exception e) {
+            logger.error("View render exception : {}", e);
```
exception 은 argument로 처리하지 않아도 됩니다!

```
logger.error(“View render exception”, e);
```

kingbbode commented on src/main/java/core/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -1,66 +1,79 @@
-public class AnnotationHandlerMapping {
+public class AnnotationHandlerMapping implements HandlerMapping {
 
+    private static final String DEFAULT_PACKAGE = "next.controller";
```
코어 패키지 안으로 어플리케이션 속성이 (“next.controller”) 들어왔네요. 어떻게 하면 코어에게 의존없이 전달할 수 있을지 고민해보는 것도 좋을 것 같아요.

kingbbode reviewed (changes requested):
간단한 피드백 남겼습니다 :)

(+) core 패키지에 작성된 웹 프레임워크에서도 요청과 응답에 대한 검증이 필요해보입니다.

juyoungyoo committed:
refactor #step3 Exception 피드백 반영

- handleException 잘못된 import 수정
- NotFoundServletException 시 상태코드 404 추가
- error log 수정

juyoungyoo committed:
refactor #step3 package(next.controller) 외부에서 주입으로 변경

juyoungyoo committed:
style #step3 jspView 클래스 명 변경

juyoungyoo committed:
refactor #step3 NotFoundException 리팩토링 및 테스트 코드 추가

juyoungyoo committed:
style #step3

- HandlerMapping 메소드 명 변경
- 로그, 코드 리팩토링
- ControllerScanner 테스트 코드 추가

juyoungyoo committed:
refactor #step3 HandlerException 오타 수정

juyoungyoo committed:
style #step3 접근제어자 변경 및 메소드 위치 수정

juyoungyoo committed:
feat #step3 viewGenerator 생성 및 패키지 변경

juyoungyoo committed:
feat #step3 annotationMapping 패키지 외부주입

juyoungyoo committed:
style step3 패키지 변경

juyoungyoo committed:
test step3 테스트 코드 수정

juyoungyoo committed:
style step3 코드 정리

juyoungyoo commented:
안녕하세요 용근님
피드백 적용했습니다. 

1. 코어에 어플리케이션 속성 제거
2. viewGenerator 구현
3. Exception 처리 

감사합니다!

kingbbode commented on src/main/java/core/mvc/DispatcherServlet.java:
```diff
@@ -0,0 +1,70 @@
+                    .filter(handlerMapping -> handlerMapping.supports(req))
+                    .findFirst()
+                    .map(serve(req, resp))
+                    .orElseThrow(NotFoundServletException::new);
```
Exception 을 처리하는 기준은 서로 다른 Layer 에서 이상 상황을 전달하는 방식이에요.
그 Layer 혹은 클래스, 메서드의 역할에 있어서 예외라고 판단되는 상황을 전달한다는 말이에요. Layer 라는 것은 우리가 알고 있는 Layer일 수도 있고 작성하고 계신 프레임워크에서 새롭게 정의한 Layer일 수도 있어요.

보통 흔한 방식으로 우리는 Layer를 Class 혹은 더 나아가면 package로도 분리를 하곤 합니다.

위 구성한 예외를 스스로 던지고 스스로 처리하는 것은 다르게 생각하면 이미 이 메서드가 여러 레이어에 걸친 작업을 수행한다고 볼 수도 있고, 불필요하게 레이어를 분리한다고까지 생각될 수도 있어요.

스트림도 물론 좋지만 위 로직은 스트림 없이 처리되어야 더 명확하게 처리될 수 있을 것 같아요.

kingbbode commented on src/main/java/core/mvc/tobe/view/ViewGenerator.java:
```diff
@@ -0,0 +1,19 @@
+
+import core.mvc.View;
+
+public class ViewGenerator {
```
이미 알고 계실 수도 있지만, 이와 같은 개념이 스프링 프레임워크에서는 ViewResolver 와 같은 네이밍으로 사용되고 있습니다 :)

kingbbode reviewed (approved):
피드백이 잘 반영되었네요 :) 
피드백 관련하여 몇 가지 코멘트를 추가로 작성하였습니다.
수고하셨습니다!

kingbbode commented:
다른 분들에게 남겼던 동일한 피드백을 추가로 남겨봅니다. 프레임워크를 구성함에 아래 내용을 꼭 생각해보셨으면 합니다 :)

1. 현재 구현하신 DispatcherServlet 은 마치 DispatcherServlet 안에 프레임워크가 구현되어가고 있는 구조인 것 같아요. 프레임워크라고 하면 서블릿 컨테이너의 라이프사이클과는 별게로 프레임워크만의 라이프사이클이 있어야하지 않을까요? DispatcherServlet이 만들고 계신 프레임워크의 일부과 될 수 있는 구조가 되면 좋을 것 같습니다.

2. 추가로 혼돈이 오실만한부분은 어디까지가 서블릿 컨테이너의 역할이고, 어디까지가 프레임워크의 역할인가 일 것 같은데요. 이 부분을 고민해보시면 좋을 것 같습니다 :)

3. 수많은 클래스들이 만들어지면서 점점 관계와 생성, 소멸에 대한 시점이 어려워지고 있네요 :) 라이프사이클을 책임지는 Context를 만들어서 정리해보시면 좋을 것 같습니다 :)
